babel/numbers.py

```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
+        number string that is strictly following CLDR pattern definitions.
+        """
+        # Manipulates decimal instances only.
         if not isinstance(value, decimal.Decimal):
             value = decimal.Decimal(str(value))
+
+        # Scale the number.
         value = value.scaleb(self.scale)
+
+        # Separate the absolute value from its sign.
         is_negative = int(value.is_signed())
-        if self.exp_prec:  # Scientific notation
+        value = abs(value).normalize()
+
+        # Prepare scientific notation metadata.
+        if self.exp_prec:
+
+            # Normalize value to only have one lead digit.
             exp = value.adjusted()
-            value = abs(value)
-            # Minimum number of integer digits
-            if self.int_prec[0] == self.int_prec[1]:
-                exp -= self.int_prec[0] - 1
-            # Exponent grouping
-            elif self.int_prec[1]:
-                exp = int(exp / self.int_prec[1]) * self.int_prec[1]
-            if exp < 0:
-                value = value * 10**(-exp)
-            else:
-                value = value / 10**exp
+            value = value * self.quantum(exp)
+            assert value.adjusted() == 0
+
+            # Shift exponent and value by the minimum number of leading digits
+            # imposed by the rendering pattern. And always make that number
+            # greater or equal to 1.
+            lead_shift = max([1, min(self.int_prec)]) - 1
+            exp = exp - lead_shift
+            value = value * self.quantum(-lead_shift)
+
+            # Get exponent sign symbol.
             exp_sign = ''
             if exp < 0:
                 exp_sign = get_minus_sign_symbol(locale)
             elif self.exp_plus:
                 exp_sign = get_plus_sign_symbol(locale)
+
+            # Normalize exponent value now that we have the sign.
             exp = abs(exp)
-            number = u'%s%s%s%s' % \
-                (self._format_significant(value, frac_prec[0], frac_prec[1]),
-                 get_exponential_symbol(locale), exp_sign,
-                 self._format_int(str(exp), self.exp_prec[0],
-                                  self.exp_prec[1], locale))
-        elif '@' in self.pattern:  # Is it a siginificant digits pattern?
-            text = self._format_significant(abs(value),
+
+        # Adjust the precision of the fractionnal part.
+        frac_prec = self.frac_prec
+        # Force precision to the currency's.
+        if currency and currency_digits:
+            frac_prec = (get_currency_precision(currency), ) * 2
+
+        # Disabled decimal quantization will ensure no loss of precision when
+        # rendering the number.
+        if not decimal_quantization:
+            # Bump decimal precision to the natural precision of the number if
+            # it exceeds the one we're about to use.
+            frac_prec = (frac_prec[0], max([frac_prec[1], self.precision(value)]))
```

Comment:
I think we are ready to see that split now.  And I'm going to insist more on separating refactoring from the real decimal quantization optional behavior code.  It will help potential future bisections.